docs/specs/alcproxies.md

```diff
@@ -0,0 +1,251 @@
+* Don't support non-primitive objects being passed to and from the ALCs
+This isn't recommended since it limits use of the target ALC by quite a lot, but for a quick and dirty version, it could work.
+* For each non-primitive object, generate another proxy that passes back to the user ALC instead of the reference to the normal object. When we unload, we can then remove this proxy as well, keeping the isolation barrier intact.
+This is very similar to how AppDomains handle MarshallByRef objects, and it keeps the functionality the same, but also adds a lot more overhead from creating the proxies during runtime for every non-primitive object.
+* Keep everything call-by-value, by making sure any `ServerObject` returns to the user ALC is the metadata of any object, that gets reconstructed when returning to `ClientObject`
+This option is nice since it works for both inter-process and in-process communication. However, having every object only able to return by value means that we'll be limited to making calls changing objects to a single ALC, and the multiple copies of an object may increase space and cause performance issues.
+
+For my planned implementation, I'll be trying to implement call-by-value, since it keeps the design consistent for all types (primitive and non-primitive) and for in-proc vs cross-proc. However, due to the design for extensibility, it's possible that anyone who wants to change the API can implement any of these options in their own way, by changing the interactions of `ClientObject` and `ServerObject`. If someone wants to limit interactions between plugins and the main context to primitives only, they're welcome to do so with this design.
+
+#### Generic Types
+For generics, the API should probably be able to allow users to pass in a list of types "in order" that are being used for the particular instance of an object they are creating. 
+
+`DispatchProxy` should be OK to use generic types, but it may be an issue funneling these types through whatever server structure is created. 
+
+For constraints on certain generics, the compiler should pick up any issues with constraints from the given interface, but may not from constraints for the TargetObject, and will throw an error.
+
+#### Performance
+No matter the performance of this API, by design it won't have the broader performance issues that AppDomains had in .Net Framework, since they work very similarly to normal classes. However, there may be some overhead and performance issues on any objects being contacted by the proxies, which should be investigated if they come up.
+
+Some hard points of performance that probably will need to be looked at:
+* Client/server calls on in-process ALCs
+* Multiple uses of `Reflection.Invoke` possibly being called
+
+#### Special Language Features
+Language features such as `in`, `ref`, and `out` should be investigated further, but we probably don't want them since the design is based around keeping everything call-by-value.
```

Comment:
We do want them.
It is fair to say that it is out of scope for this initial feasibility prototype.